text/0000-project-ffi-unwind.md

```diff
@@ -0,0 +1,211 @@
+
+* The group does not intend to consider mechanisms to enable "interop"
+  between Rust panics and exceptions from other languges. For example,
+  we do not intend to permit Rust code to catch C++ exceptions, though
+  we will have to consider what happens when a C++ exception unwinds
+  past a `catch_unwind` boundary.
+
+### Constraints and considerations
+
+In its work, the project-group should consider various constraints and
+considerations:
+
+* The possibility that C++ may adopt new unwinding mechanisms in the future.
+* The possibility that Rust may alter its unwinding mechanism in the future --
+  in particular, the project group must not propose a design that would
+  constrain Rust's unwinding implementation.
+
+### Participation in the project group
+
+Like any Rust group, the FFI-unwind project group intends to operate
+in a public and open fashion and welcomes participation. Visit the
+[repository][ffi-unwind project] for more details.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The adoption of project groups for major language design efforts is a change
+  in the status quo. We believe that this change will be an improvement over
+  the current RFC-centric process, but we should be wary of unintended
+  consequences of from such a change.
+* [Stabilization of "TBD" features][stabilizing-tbd] may be surprising or
+  confusing to users, and it will encourage reliance on (some) unspecified
+  behavior.
+
+# Prior art
+[prior-art]: #prior-art
+
+Although the term "project group" is new, some existing efforts, such as the
+Unsafe Code Guidelines effort and the work around defining const evaluation,
+were organized in a similar fashion.
+
+In addition to the [blog post Niko Matsakis][shepherds-3.0] about
+shepherding, James Munns wrote a [more formal shepherding
+proposal][shepherding-3.1].
+
+The [governance WG][governance-wg] and [lang-team meta working
+group][lang-meta-wg] were both formed at least in part to improve the process
+for large-scale design efforts. One existing proposal is for ["staged
+RFCs"][staged-rfc]; this may be considered a precursor to the current
+"shepherded project group" proposal.
+
+
+# Unresolved questions and Future possibilities
+[unresolved-questions]: #unresolved-questions
+
+Since this RFC merely formalizes the creation of the project group, it
+intentionally leaves all technical details within the project's scope
+unresolved.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+The project group will start with a fairly [limited scope][charter], but if the
+initial effort to design and stabilize a safe cross-language unwinding feature
+on a limited set of platforms goes well, there are many related areas of
+potential exploration. Three noteworthy examples are:
+
+* Catching foreign unwinding (e.g. Rust catching C++ exceptions, or C++
+  catching Rust `panic`s)
+* Defining subtype relationships among `fn`s using ABIs with different `unwind`
```

Comment:
Is this about *top level* coercions? I don't think anyone has proposed that e.g. `x: Vec<extern "C" fn(...)> as Vec<extern "C unwind" fn(...)>` be a legal transformation.

Reply:
I don't know what "top level" means here, and I don't actually know the full rules for covariance in Rust (e.g. would making `extern "C" fn as extern "C unwind" fn` automatically imply that your example would also be valid?).

As far as I can remember, previous discussions on ABI-to-ABI coercions involving `unwind` have mostly been limited to agreeing that for simplicity's sake we shouldn't even worry about them yet, since it's always possible to wrap a function pointer in a lambda with a different ABI. I don't think any additional details or constraints are warranted in this document.

Reply:
```rust
fn main() {
    let f: fn() = main;
    let uf: unsafe fn() = f; // OK.
    //~^ Top level coercion.

    let vf: &fn() = &f;
    let vuf: &unsafe fn() = vf; // ERROR
    //~^ Subtyping. A reference `&'a T` is covariant wrt. `T`.
    // That means that if `X <: Y` we can take a value `x: &'a X`
    // and use it where `&'a Y` is expected.
    // However, no subtyping relationship holds between `fn() <: unsafe fn()`.
    // Therefore, `&'a fn()` may not be used where `&'a unsafe fn` is wanted.
}
```

> I don't think any additional details or constraints are warranted in this document.

Sure, but I think it would be more realistic if the document said:

```suggestion
* Defining coercions among `fn`s using ABIs with different `unwind`
```

Reply:
> Is this about top level coercions? I don't think anyone has proposed that e.g. x: Vec<extern "C" fn(...)> as Vec<extern "C unwind" fn(...)> be a legal transformation.

@Centril somebody did: https://github.com/rust-lang/project-ffi-unwind/pull/9/files#diff-1b91cd2b717d4c4e383aec41a4ad206eR304

Reply:
@Centril Thanks for the clarification. I didn't realize there were cases where implicit coercion could be used without a subtyping relationship. I can't say I fully understand that yet, but I've changed the wording appropriately.